README.md
# Lazy Page View package

[![pub package](https://img.shields.io/pub/v/lazy_page_view.svg)](https://pub.dev/packages/lazy_page_view)

Provides a wrapper for the default PageView to allow for efficient page by page loading of widgets.
Useful for asynchronous database queries, API requests or other expensive operations.

#### This package is also highly useful, if the total page amount and initial index is unknown, but the PageView should still be constrained at the start and end of the pages.

This package gives interfaces to allow for the loading of data with full type and null safety, and
provides a simple way to make next / previous pages data load dependent on the current pages data.

Using the exposed members of the controller, pages can also be loaded / reloaded manually, and the current page index can be accessed (relative to the initial page index).

## Features

- [x] Lazy loading of pages
- [x] Loading of next / previous pages data dependent on current page data
- [x] Full type and null safety
- [x] Relative page index based on initial page as index 0
- [x] Customisable loading widget

## Getting started

1. Prepare your widget to display the LazyPageView
2. Create your methods for loading the initial page data, and each of the next / previous pages data based on the current page data
3. Create your LazyPageView widget, passing in the required parameters
4. [Optional] Use a LazyPageController to control the LazyPageView, giving you access to the currentPage index and a page change listener

## Usage

```dart
import 'package:flutter/material.dart';
import 'package:lazy_page_view/completion.dart';
import 'package:lazy_page_view/lazy_page_controller.dart';
import 'package:lazy_page_view/lazy_page_view.dart';

import 'database_service.dart';
import 'event.dart';

class ExampleLazyPageViewImplementation extends StatefulWidget {
  const ExampleLazyPageViewImplementation({super.key});

  @override
  State<ExampleLazyPageViewImplementation> createState() => _ExampleLazyPageViewImplementationState();
}

class _ExampleLazyPageViewImplementationState extends State<ExampleLazyPageViewImplementation> {
  LazyPageController<Event> _lazyPageController = LazyPageController<Event>();
  
  Future<Event?> getNextDatabaseEvent(Event currentEvent) async {
    // Load the next event or null if there are no more events from the database, based on the current event
    return await DatabaseService.getNextEvent(currentEvent);
  }
  
  Future<Event?> getPreviousDatabaseEvent(Event currentEvent) async {
      // Load the previous event or null if there are no more events from the database, based on the current event
      return await DatabaseService.getPreviousEvent(currentEvent);
  }
  
  Future<Event> loadInitialEvent() async {
    // Load the initial event from the database (for example, the next event to occur in the future)
    return await DatabaseService.getInitialEvent();
  }
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: LazyPageView<Event>(
        controller: _lazyPageController,
        loadInitial: loadInitialEvent,
        loadNext: getNextDatabaseEvent,
        loadPrevious: getPreviousDatabaseEvent, 
        pageBuilder: (BuildContext context, Event event) {
          return EventPage(event);
        },
        placeHolder: const Center(child: CircularProgressIndicator()),
        onPageChanged: (Event event) {
          print('Page changed to: $event');
        },
      ),
    );
  }
}

```


CHANGELOG.md
## 1.0.0

Initial release, with working example implemented in a production use app.

lib/completion.dart
/// The state of a completion.
enum CompletionState {
  /// The future is currently loading.
  loading,

  /// The future has completed with a value or null.
  loaded,

  /// The future has completed with an error.
  error,
}

/// A wrapper around a future that allows you to check the state of the future
/// and get the value of the future at any time.
class Completion<T> {
  /// The future that this completion wraps.
  Future<T?> future;

  /// The current state of the future.
  CompletionState state = CompletionState.loading;

  /// The value of the future.
  T? value;

  /// The error of the future, if any.
  dynamic error;

  /// The stack trace of the future, if any.
  StackTrace? stackTrace;

  Completion(this.future) {
    future.then((value) {
      this.value = value;
      state = CompletionState.loaded;
    }).catchError((e, s) {
      this.error = e;
      this.stackTrace = s;
      state = CompletionState.error;

      throw e;
    });
  }

  /// Attempt to get the value of the future.
  ///
  /// If the future has not yet completed, this will throw an exception.
  /// If the future has completed with an error, this will also throw an exception.
  T? get() {
    if (state == CompletionState.loading) {
      throw Exception('Cannot get value of a loading completion');
    } else if (state == CompletionState.error) {
      throw Exception('Cannot get value of a errored completion');
    }
    return value;
  }

  bool get isLoaded => state == CompletionState.loaded;
  bool get isLoading => state == CompletionState.loading;
  bool get isError => state == CompletionState.error;
}


lib/lazy_pageview.dart
library lazy_pageview;

import 'package:flutter/material.dart';
import 'package:lazy_pageview/lazy_pageview_controller.dart';

import 'completion.dart';

/// A PageView that loads pages lazily.
///
/// It should be used, if either of the following scenarios is true:
/// - The page data is expensive to load (This widget will build only the pages that are visible)
/// - The size of the page view is dynamic and might change at any time with no "real" index existing
/// - Creation of pages are relative and not absolute, so the data depends on each other
class LazyPageView<T> extends StatefulWidget {
  const LazyPageView({
    super.key,
    required this.loadInitial,
    required this.loadPrevious,
    required this.loadNext,
    required this.pageBuilder,
    this.placeholder = const Center(child: CircularProgressIndicator()),
    this.onPageChanged,
    this.onLeftEndReached,
    this.onRightEndReached,
    this.onLeftEndUnreached,
    this.onRightEndUnreached,
    this.controller,
  });

  final Future<T?> Function() loadInitial;
  final Future<T?> Function(T current) loadPrevious;
  final Future<T?> Function(T current) loadNext;
  final Widget Function(BuildContext context, T data) pageBuilder;
  final Widget placeholder;

  final LazyPageViewController<T>? controller;

  final void Function(T data)? onPageChanged;
  final void Function(T data)? onLeftEndReached;
  final void Function(T data)? onRightEndReached;
  final void Function(T data)? onLeftEndUnreached;
  final void Function(T data)? onRightEndUnreached;

  @override
  State<LazyPageView> createState() => _LazyPageViewState<T>();
}

class _LazyPageViewState<T> extends State<LazyPageView> {
  PageController pageController = PageController(keepPage: false, initialPage: 10000);
  double lastX = 0;

  late LazyPageViewController controller;

  @override
  void dispose() {
    pageController.dispose();
    if (widget.controller == null) {
      controller.dispose();
    }
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    if (widget.controller != null) {
      controller = widget.controller!;
    } else {
      controller = LazyPageViewController<T>();
    }

    controller.currentPageData = Completion<T?>(loadInitially());
  }

  Future<T?> loadInitially() async {
    T? data = await widget.loadInitial();
    controller.currentPageData = Completion<T?>(Future.value(data));
    controller.previousPageData = Completion<T?>(loadPrevious(true));
    controller.nextPageData = Completion<T?>(loadNext(true));
    if (mounted) setState(() {});
    return data;
  }

  Future<T?> loadNext([bool initial = false]) async {
    T? data = await widget.loadNext(controller.currentPageData.get());
    if (data == null) {
      controller.rightEndReached = true;
      if (widget.onRightEndReached != null && controller.currentPageData.isLoaded) {
        widget.onRightEndReached!(controller.currentPageData.get());
      }
    } else {
      controller.rightEndReached = false;
      if (widget.onRightEndUnreached != null && !initial && controller.currentPageData.isLoaded) {
        widget.onRightEndUnreached!(controller.currentPageData.get());
      }
    }
    return data;
  }

  Future<T?> loadPrevious([bool initial = false]) async {
    T? data = await widget.loadPrevious(controller.currentPageData.get());
    if (data == null) {
      controller.leftEndReached = true;
      if (widget.onLeftEndReached != null && controller.currentPageData.isLoaded) {
        widget.onLeftEndReached!(controller.currentPageData.get());
      }
    } else {
      controller.leftEndReached = false;
      if (widget.onLeftEndUnreached != null && !initial && controller.currentPageData.isLoaded) {
        widget.onLeftEndUnreached!(controller.currentPageData.get());
      }
    }
    return data;
  }

  @override
  Widget build(BuildContext context) {
    return Listener(
      onPointerMove: (event) {
        double x = event.position.dx;
        if (x > lastX && controller.leftEndReached) {
          pageController.animateToPage(controller.pageViewIndex, duration: const Duration(milliseconds: 100), curve: Curves.ease);
        }
        if (x < lastX && controller.rightEndReached) {
          pageController.animateToPage(controller.pageViewIndex, duration: const Duration(milliseconds: 100), curve: Curves.ease);
        }

        lastX = x;
      },
      child: PageView.builder(
          controller: pageController,
          onPageChanged: (page) async {
            if (page == controller.pageViewIndex) return;
            if (page == controller.pageViewIndex + 1) {
              controller.pageViewIndex = page;
              if (!controller.nextPageData.isLoaded) return;
              if (mounted) {
                setState(() {
                  controller.leftEndReached = false;

                  if (!controller.rightEndReached) {
                    controller.next(loadNext());

                    if (widget.onPageChanged != null && controller.currentPageData.isLoaded) {
                      widget.onPageChanged!(controller.currentPageData.get());
                    }
                  } else {
                    controller.pageViewIndex = page - 1;
                  }
                });
              }
            } else if (page == controller.pageViewIndex - 1) {
              controller.pageViewIndex = page;
              if (!controller.previousPageData.isLoaded) return;
              if (mounted) {
                setState(() {
                  controller.rightEndReached = false;

                  if (!controller.leftEndReached) {
                    controller.previous(loadPrevious());

                    if (widget.onPageChanged != null && controller.currentPageData.isLoaded) {
                      widget.onPageChanged!(controller.currentPageData.get());
                    }
                  } else {
                    controller.pageViewIndex = page + 1;
                  }
                });
              }
            }
          },
          itemBuilder: (context, index) {
            if (controller.currentPageData.isLoading) return widget.placeholder;

            if (controller.pageViewIndex == index - 1 && controller.nextPageData.isLoaded && !controller.rightEndReached) {
              return widget.pageBuilder(context, controller.nextPageData.get());
            }
            if (controller.pageViewIndex == index && controller.currentPageData.isLoaded) {
              return widget.pageBuilder(context, controller.currentPageData.get());
            }
            if (controller.pageViewIndex == index + 1 && controller.previousPageData.isLoaded && !controller.leftEndReached) {
              return widget.pageBuilder(context, controller.previousPageData.get());
            }

            if (controller.pageViewIndex != index) {
              if (controller.leftEndReached || controller.rightEndReached) {
                return const SizedBox();
              }
            }

            WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
              if (controller.leftEndReached) {
                pageController.animateToPage(controller.pageViewIndex + 1, duration: const Duration(milliseconds: 200), curve: Curves.ease);
              } else if (controller.rightEndReached) {
                pageController.animateToPage(controller.pageViewIndex - 1, duration: const Duration(milliseconds: 200), curve: Curves.ease);
              }
            });

            return widget.placeholder;
          }),
    );
  }
}


lib/lazy_page_controller.dart
import 'package:flutter/cupertino.dart';
import 'package:lazy_page_view/completion.dart';
import 'package:lazy_page_view/lazy_page_view.dart';

class LazyPageController<T> extends ChangeNotifier {
  static const int initialPage = 10000;

  late Completion<T?> currentPageData;
  late Completion<T?> previousPageData;
  late Completion<T?> nextPageData;

  bool leftEndReached = false;
  bool rightEndReached = false;

  int pageViewIndex = initialPage;

  /// The current page index, relative to the initial page.
  ///
  /// For each call of [next] manually or by the [LazyPageView], the index is incremented by one.
  /// For each call of [previous] manually or by the [LazyPageView], the index is decremented by one.
  int get currentPage {
    return pageViewIndex - initialPage;
  }

  void next(Future<T?> Function() future) {
    previousPageData = currentPageData;
    currentPageData = nextPageData;
    nextPageData = Completion<T?>(future());

    notifyListeners();
  }

  void previous(Future<T?> Function() future) {
    nextPageData = currentPageData;
    currentPageData = previousPageData;
    previousPageData = Completion<T?>(future());

    notifyListeners();
  }
}


lib/lazy_pageview_controller.dart
import 'package:flutter/cupertino.dart';
import 'package:lazy_pageview/completion.dart';

class LazyPageViewController<T> extends ChangeNotifier {
  static const int initialPage = 10000;

  late Completion<T?> currentPageData;
  late Completion<T?> previousPageData;
  late Completion<T?> nextPageData;

  bool leftEndReached = false;
  bool rightEndReached = false;

  int pageViewIndex = initialPage;
  int get currentPage {
    return pageViewIndex - initialPage;
  }

  void next(Future<T?> future) {
    previousPageData = currentPageData;
    currentPageData = nextPageData;
    nextPageData = Completion<T?>(future);

    notifyListeners();
  }

  void previous(Future<T?> future) {
    nextPageData = currentPageData;
    currentPageData = previousPageData;
    previousPageData = Completion<T?>(future);

    notifyListeners();
  }
}


lib/lazy_page_view.dart
library lazy_page_view;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:lazy_page_view/lazy_page_controller.dart';

import 'completion.dart';

/// A PageView that loads pages lazily.
///
/// It should be used, if either of the following scenarios is true:
/// - The page data is expensive to load (This widget will build only the pages that are visible)
/// - The size of the page view is dynamic and might change at any time with no "real" index existing
/// - Creation of pages are relative and not absolute, so the data depends on each other
class LazyPageView<T> extends StatefulWidget {
  const LazyPageView({
    super.key,
    required this.loadInitial,
    required this.loadPrevious,
    required this.loadNext,
    required this.pageBuilder,
    this.placeholder = const Center(child: CircularProgressIndicator()),
    this.onPageChanged,
    this.onLeftEndReached,
    this.onRightEndReached,
    this.onLeftEndUnreached,
    this.onRightEndUnreached,
    this.controller,
    this.allowImplicitScrolling,
    this.clipBehavior,
    this.dragStartBehavior,
    this.scrollDirection,
  });

  /// A function that loads the initial pages data
  final Future<T?> Function() loadInitial;

  /// A function that loads the previous pages data depending on the current page data
  final Future<T?> Function(T current) loadPrevious;

  /// A function that loads the next pages data depending on the current page data
  final Future<T?> Function(T current) loadNext;

  /// A builder that builds the page from the data
  ///
  /// ## NOTE: if [pageSnapping] is false, the itemBuilder must return widgets larger or equal to the viewport!
  final Widget Function(BuildContext context, T data) pageBuilder;

  /// A widget to display while the page is loading
  final Widget placeholder;

  /// An object that can be used to control how the [LazyPageView] functions
  final LazyPageController<T>? controller;

  /// Called whenever the page in the center of the viewport changes.
  final void Function(T data)? onPageChanged;

  /// Called whenever the loadPrevious function returns null.
  final void Function(T data)? onLeftEndReached;

  /// Called whenever the loadNext function returns null.
  final void Function(T data)? onRightEndReached;

  /// Called whenever the loadPrevious function returns a value after previously having returned null.
  final void Function(T data)? onLeftEndUnreached;

  /// Called whenever the loadNext function returns a value after previously having returned null.
  final void Function(T data)? onRightEndUnreached;

  /// Controls whether the widget's pages will respond to
  /// [RenderObject.showOnScreen], which will allow for implicit accessibility
  /// scrolling.
  ///
  /// With this flag set to false, when accessibility focus reaches the end of
  /// the current page and the user attempts to move it to the next element, the
  /// focus will traverse to the next widget outside of the page view.
  ///
  /// With this flag set to true, when accessibility focus reaches the end of
  /// the current page and user attempts to move it to the next element, focus
  /// will traverse to the next page in the page view.
  /// {@macro flutter.material.Material.clipBehavior}
  final bool? allowImplicitScrolling;

  /// {@macro flutter.material.Material.clipBehavior}
  ///
  /// Defaults to [Clip.hardEdge].
  final Clip? clipBehavior;

  /// {@macro flutter.widgets.scrollable.dragStartBehavior}
  final DragStartBehavior? dragStartBehavior;

  /// The [Axis] along which the scroll view's offset increases with each page.
  ///
  /// For the direction in which active scrolling may be occurring, see
  /// [ScrollDirection].
  ///
  /// Defaults to [Axis.horizontal].
  final Axis? scrollDirection;

  @override
  State<LazyPageView<T>> createState() => _LazyPageViewState<T>();
}

class _LazyPageViewState<T> extends State<LazyPageView<T>> {
  PageController pageController = PageController(keepPage: false, initialPage: 10000);
  double lastPos = 0;

  late LazyPageController controller;

  @override
  void dispose() {
    pageController.dispose();
    if (widget.controller == null) {
      controller.dispose();
    }
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    if (widget.controller != null) {
      controller = widget.controller!;
    } else {
      controller = LazyPageController<T>();
    }

    controller.currentPageData = Completion<T?>(loadInitially());
  }

  Future<T?> loadInitially() async {
    T? data = await widget.loadInitial();
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      if (mounted) {
        controller.previousPageData = Completion<T?>(loadPrevious(true));
        controller.nextPageData = Completion<T?>(loadNext(true));
        if (mounted) setState(() {});
      }
    });
    if (mounted) setState(() {});
    return data;
  }

  Future<T?> loadNext([bool initial = false]) async {
    T? a = controller.currentPageData.get();
    Future<T?> b = widget.loadNext(a as T);
    T? data = await b;
    if (data == null) {
      controller.rightEndReached = true;
      if (widget.onRightEndReached != null && controller.currentPageData.isLoaded) {
        widget.onRightEndReached!(controller.currentPageData.get());
      }
    } else {
      controller.rightEndReached = false;
      if (widget.onRightEndUnreached != null && !initial && controller.currentPageData.isLoaded) {
        widget.onRightEndUnreached!(controller.currentPageData.get());
      }
    }
    return data;
  }

  Future<T?> loadPrevious([bool initial = false]) async {
    T? a = controller.currentPageData.get();
    Future<T?> b = widget.loadPrevious(a as T);
    T? data = await b;
    if (data == null) {
      controller.leftEndReached = true;
      if (widget.onLeftEndReached != null && controller.currentPageData.isLoaded) {
        widget.onLeftEndReached!(controller.currentPageData.get());
      }
    } else {
      controller.leftEndReached = false;
      if (widget.onLeftEndUnreached != null && !initial && controller.currentPageData.isLoaded) {
        widget.onLeftEndUnreached!(controller.currentPageData.get());
      }
    }
    return data;
  }

  @override
  Widget build(BuildContext context) {
    return PageView.builder(
        allowImplicitScrolling: widget.allowImplicitScrolling ?? false,
        clipBehavior: widget.clipBehavior ?? Clip.hardEdge,
        dragStartBehavior: widget.dragStartBehavior ?? DragStartBehavior.start,
        reverse: false,
        scrollDirection: widget.scrollDirection ?? Axis.horizontal,
        controller: pageController,
        onPageChanged: (page) async {
          if (page == controller.pageViewIndex) return;
          if (page == controller.pageViewIndex + 1) {
            controller.pageViewIndex = page;
            if (!controller.nextPageData.isLoaded) return;
            if (mounted) {
              setState(() {
                controller.leftEndReached = false;

                if (!controller.rightEndReached) {
                  controller.next(loadNext);

                  if (widget.onPageChanged != null && controller.currentPageData.isLoaded) {
                    widget.onPageChanged!(controller.currentPageData.get());
                  }
                } else {
                  controller.pageViewIndex = page - 1;
                }
              });
            }
          } else if (page == controller.pageViewIndex - 1) {
            controller.pageViewIndex = page;
            if (!controller.previousPageData.isLoaded) return;
            if (mounted) {
              setState(() {
                controller.rightEndReached = false;

                if (!controller.leftEndReached) {
                  controller.previous(loadPrevious);

                  if (widget.onPageChanged != null && controller.currentPageData.isLoaded) {
                    widget.onPageChanged!(controller.currentPageData.get());
                  }
                } else {
                  controller.pageViewIndex = page + 1;
                }
              });
            }
          }
        },
        itemBuilder: (context, index) {
          if (controller.currentPageData.isLoading) return widget.placeholder;

          if (controller.pageViewIndex == index - 1 && controller.nextPageData.isLoaded && !controller.rightEndReached) {
            return widget.pageBuilder(context, controller.nextPageData.get());
          }
          if (controller.pageViewIndex == index && controller.currentPageData.isLoaded) {
            return widget.pageBuilder(context, controller.currentPageData.get());
          }
          if (controller.pageViewIndex == index + 1 && controller.previousPageData.isLoaded && !controller.leftEndReached) {
            return widget.pageBuilder(context, controller.previousPageData.get());
          }

          if (controller.pageViewIndex != index) {
            if (controller.leftEndReached || controller.rightEndReached) {
              WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
                if (controller.leftEndReached) {
                  pageController.animateToPage(controller.pageViewIndex, duration: const Duration(milliseconds: 200), curve: Curves.ease);
                } else if (controller.rightEndReached) {
                  pageController.animateToPage(controller.pageViewIndex, duration: const Duration(milliseconds: 200), curve: Curves.ease);
                }
              });
              return const SizedBox();
            }
          }

          return widget.placeholder;
        });
  }
}
